Allow binding dig, place actions to keys; remove LMB/RMB hardcoding

ClobberXD opened the pull request:
[Adopted from #6537]

- Remove hard-coded association of dig and place actions, with LMB and RMB respectively.
  - These actions are still bound to LMB and RMB by default.
- Allow binding actions to mouse buttons.
- `ObjectRef:get_player_control` now returns `dig` and `place` fields. Note: The old `LMB` and `RMB` fields still exist alongside `dig` and `place`, to ensure backwards compatibility.

Tested, works. This PR is ready for review.

#### Video demonstration

https://youtu.be/0Ca7nEMiq_U

#### How to test

- Suggested keymap to test this PR:
  ```
  keymap_place = KEY_KEY_L
  keymap_dig = KEY_KEY_X
  keymap_inventory = KEY_LBUTTON
  keymap_jump = KEY_RBUTTON
  ```
- Tweak `repeat_place_interval`, and test it by selecting a place-able item (e.g. a node), pointing at a node, and holding down the "place" keybind.

Due to code changes around keypress detection, it's advisable to also test various digging, placing, use-ing, and secondary_use-ing scenarios, and carefully observe that nothing has broken. Note that this PR doesn't modify any part of the actual digging and placing, even in a subtle manner. So if a change is noticeable, please do let me know.

ClobberXD commented:
Rebased and fixed merge conflicts. But I'm not really happy with the choice of words used for both LMB (`DIG`) and RMB (`PLACE`) - LMB and RMB is also used for punching and misc. interacting with world respectively. `USE` and `SECONDARY_USE` would suit them better, IMHO. Thoughts?

jastevenson303 commented:
Use reminds me of "aux1" and secondary_use reminds me of on_secondary_use
node def value.  But if that doesn't matter, USE and SECONDARY_USE sound
good to me.  Could also go with "add" and "subtract"; but this could turn
into a lot of thought over a somewhat inconsequential detail.

On Sat, Dec 1, 2018 at 11:04 AM ClobberXD <notifications@github.com> wrote:

> Rebased and fixed merge conflicts. But I'm not really happy with the
> choice of words used for both LMB (DIG) and RMB (PLACE) - LMB and RMB is
> also used for punching and misc. interacting with world respectively. USE
> and SECONDARY_USE would suit them better, IMHO. Thoughts?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/7924#issuecomment-443436535>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWXKYntyPCPQnpbtP7IB1W-FZktuNsJYks5u0qiMgaJpZM4Y88Ts>
> .
>

ClobberXD commented:
> Use reminds me of "aux1" and secondary_use reminds me of on_secondary_use node def value. But if that doesn't matter, USE and SECONDARY_USE sound good to me. Could also go with "add" and "subtract"; but this could turn into a lot of thought over a somewhat inconsequential detail.

Yes, but all references to that key now use `aux1` only, and I assume it's safe to name LMB `USE`. I'm not as convinced about `SECONDARY_USE` for RMB, but it's still much better than `PLACE` which is misleading.

SmallJoker commented:
> @SmallJoker could you please elaborate on this? What exactly needs to be done to fix this compat breakage?

Change
https://github.com/minetest/minetest/blob/2d05e09e6d07e4eb3a86dd13f1750c02fde9c844/src/script/lua_api/l_object.cpp#L1313-L1320

to

```C++
	lua_pushboolean(L, control.dig);
	lua_setfield(L, -2, "dig");
	lua_pushboolean(L, control.place);
	lua_setfield(L, -2, "place");
	// Legacy fields to ensure mod compatibility
	lua_pushboolean(L, control.LMB);
	lua_setfield(L, -2, "LMB");
	lua_pushboolean(L, control.RMB);
	lua_setfield(L, -2, "RMB");
```
Please also adjust the docs accordingly.

jastevenson303 commented:
fair points clobberxd.  then how about use_secondary versus secondary_use?

On Sun, Dec 2, 2018, 5:31 AM SmallJoker <notifications@github.com> wrote:

> @SmallJoker <https://github.com/SmallJoker> could you please elaborate on
> this? What exactly needs to be done to fix this compat breakage?
>
> Change
>
> https://github.com/minetest/minetest/blob/2d05e09e6d07e4eb3a86dd13f1750c02fde9c844/src/script/lua_api/l_object.cpp#L1313-L1320
>
> to
>
> 	lua_pushboolean(L, control.dig);
> 	lua_setfield(L, -2, "dig");
> 	lua_pushboolean(L, control.place);
> 	lua_setfield(L, -2, "place");
> 	// Legacy fields to ensure mod compatibility
> 	lua_pushboolean(L, control.LMB);
> 	lua_setfield(L, -2, "LMB");
> 	lua_pushboolean(L, control.RMB);
> 	lua_setfield(L, -2, "RMB");
>
> Please also adjust the docs accordingly.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/7924#issuecomment-443497157>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWXKYhzzUGoVVnrTix0pce85oq7oHB7Sks5u06wRgaJpZM4Y88Ts>
> .
>

ClobberXD commented:
`secondary_use` sounds better than `use_secondary`

jastevenson303 commented:
When right clicking air, it's called "activating."  So how about `activate` instead of `secondary_use`?

ClobberXD commented:
> When right clicking air, it's called "activating." So how about `activate` instead of `secondary_use`?

Hmm... `Press the activate button to open/close this chest` - This doesn't sound as meaningful. The reason why I'm suggesting `secondary_use` is that it's much more generic and can apply to all the actions equally well. e.g. `Press the secondary use button to open the door.`

jastevenson303 commented:
Yep, sounds good to me.  TBH I'm just excited and a little anxious so don't
mind me.

On Thu, Dec 6, 2018, 12:52 AM ClobberXD <notifications@github.com> wrote:

> When right clicking air, it's called "activating." So how about activate
> instead of secondary_use?
>
> Hmm... Press the activate button to open/close this chest - This doesn't
> sound as meaningful. The reason why I'm suggesting secondary_use is that
> it's much more generic and can apply to all the actions equally well. e.g. Press
> the secondary use button to open the door.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/7924#issuecomment-444756670>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWXKYj7rItlfr8aL7lvbjI03Mw7haNDjks5u2LCMgaJpZM4Y88Ts>
> .
>

ClobberXD commented:
> Yep, sounds good to me. TBH I'm just excited and a little anxious so don't mind me.

I can understand. :)

I just wish I can understand why there're include errors popping up in a couple of files.

lhofhansl commented:
I'm very much in favor of the idea (haven't had a chance to look at the changes, though)

ClobberXD commented:
Squashed the commits for ease of manageability.

ClobberXD commented on src/nodedef.cpp:
```diff
@@ -366,7 +366,7 @@ void ContentFeatures::reset()
 	buildable_to = false;
 	floodable = false;
-	rightclickable = true;
+	can_place_onto = true;
```
This is misleading - `can_place_onto` is *not* the same as `rightclickable`. Please advise as to what this should be changed to.

ClobberXD replied:
Never mind, reverted to `rightclickable` for now.

ClobberXD commented:
Rebased. This PR should be ready for testing.

Wuzzy2 commented:
Hmmm, I am not happy with the terms `USE` and `SECONDARY_USE` tbh.

Because leftclick's main (!) use is “dig/punch”, and rightclick's main use is “use/place”, where with “use” we mean “using the pointed node or thing”. At least this is defined that way in the Esc menu and README file. And it's mostly true. It's only that special items might subvert these definitions.

Calling the key that is primarily used for digging/punching “USE” is very confusing. And the name “SECONDARY_USE” does not make it any better.

ClobberXD commented:
> Hmmm, I am not happy with the terms USE and SECONDARY_USE tbh.
>
> Calling the key that is primarily used for digging/punching “USE” is very confusing. And the name “SECONDARY_USE” does not make it any better.

I agree that `USE` and `SECONDARY_USE` aren't the best names possible - while they're pretty generalised, they also do a bad job of actually describing what the buttons do.

> Because leftclick's main (!) use is “dig/punch”, and rightclick's main use is “use/place”, where with “use” we mean “using the pointed node or thing”. At least this is defined that way in the Esc menu and README file. And it's mostly true. It's only that special items might subvert these definitions.

So, do you suggest using "dig/punch" and "use/place" as the button names? They describe the function of these buttons quite well, but how would they be presented to the end-user the same way? I mean, isn't a single term better for this, rather than two terms with a `/` in the middle?

jastevenson303 commented:
Tested.  Did not work.  I was able to jump with right mouse button, but couldn't place or dig regardless of where those two actions were bound.

jastevenson303 commented:
Also can we please just settle on a name?  LOL

ClobberXD commented:
@jastevenson303 I haven't got around to testing this yet, so thanks for testing this. I'll look into why it doesn't work.

ClobberXD commented:
Can confirm - actions are bindable to mouse buttons and work properly, but dig and place don't work irrespective of their keybinds.

jastevenson303 commented on src/client/game.cpp:
```diff
@@ -595,15 +594,16 @@ struct GameRunData {
 	bool digging_blocked;
-	bool left_punch;
+	bool punch;
+	bool update_wielded_item_trigger;
```
It appears this line was not an addition in the original PR?  Just curious if it was taken out at some point.  Good luck, I hope you can get this to work again.  :)

ClobberXD replied:
This line seems to have already existed in the code, before the original PR, but was removed in a later commit. This would've been added during rebase. I still kinda suck at rebasing in all but the simplest of situations :)

ClobberXD commented on builtin/settingtypes.txt:
```diff
@@ -110,9 +110,9 @@ doubletap_jump (Double tap jump for fly) bool false
-repeat_rightclick_time (Rightclick repetition interval) float 0.25
+#    The time in seconds it takes between repeated block placements when holding
+#    the place block button.
+repeat_place_time (Place repetition interval) float 0.25
```
This would break existing clients in a non-violent way. i.e. `repeat_rightclick_interval` won't work anymore, but they could easily fix this by renaming it to `repeat_place_time`. Thoughts?

ClobberXD replied:
This needs attending to as well. I doubt this is a popular setting, but I'm pretty sure a good bunch of users have tweaked this to their requirements. IMO, the new name actually better describes the purpose of the setting, even if `place` is hardcoded to RMB.

SmallJoker replied:
Backwards compatibility would be nice to have.
1) Read old setting value (if it exists)
2) Set new value
3) Remove old setting line

Wuzzy2 commented:
> This would break existing clients in a non-violent way. i.e. repeat_rightclick_interval won't work anymore, but they could easily fix this by renaming it to repeat_place_time. Thoughts?

Does this work? E.g. has this been tested to actually *only* place things, and not trigger `on_rightclick` of pointed things? This is important.

ClobberXD commented on src/client/game.cpp:
```diff
@@ -3012,51 +3009,51 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
-	input->resetLeftReleased();
-	input->resetRightReleased();
+	input->joystick.clearWasKeyReleased(KeyType::MOUSE_L);
+	input->joystick.clearWasKeyReleased(KeyType::MOUSE_R);
```
Does `KeyType::MOUSE_L` and `KeyType::MOUSE_R` need to be changed to `KeyType::DIG` and `KeyType::PLACE`?

rubenwardy replied:
joysticks largely don't work in Minetest :D

I suspect it may need to be DIG and PLACE though

ClobberXD replied:
I expected that answer, but I don't want to screw it up even further. :)

Anyway, I'll see what I can do.

ClobberXD replied:
I'll leave joystick unhardcoding to another PR as I'm completely unfamiliar with joystick code at the moment. Even if I try to work on it, this PR is going to be greatly delayed yet again.

ClobberXD commented:
> Does this work? E.g. has this been tested to actually only place things, and not trigger on_rightclick of pointed things? This is important.

`repeat_rightclick_interval` just limits the number of right-clicks simulated when RMB is held down. This PR just changes the name of the setting, but the behaviour and function of this setting is the exact same. i.e. if the pointed node defines `on_rightclick`, that is triggered, and the wielded node isn't placed. e.g. toggling doors. This is how it currently works too. Maybe I misunderstood what you said?

Wuzzy2 commented:
But isn't the new name extremely misleading then? Triggering `on_rightclick` != placing a node.

Wuzzy2 commented:
Note: I agree the name `rightclick_interval` is misleading, too, if mouse buttons are bindable. Need to think of a better name.

ClobberXD commented on src/nodedef.h:
```diff
@@ -354,7 +354,7 @@ struct ContentFeatures
 	bool buildable_to;
 	// Player cannot build to these (placement prediction disabled)
-	bool rightclickable;
+	bool can_place_onto;
```
Another instance of the ironical nodedef property :)

> ```cpp
> // Player cannot build to these (placement prediction disabled)
> bool can_place_onto;
> ```

Note that this change also breaks existing mods that define nodes with the `rightclickable` property, and it's probably best to revert all changes to this nodedef property. Thoughts?

DS-Minetest replied:
Nodes can't be defined with the `rightclickable` property. This bool just says whether an `on_rightclick` field is set:
https://github.com/minetest/minetest/blob/4d7f2967172c5a358811ac1f24115b281eff2431/src/script/common/c_content.cpp#L507-L508

ClobberXD replied:
Oh, that clears things up, thanks.

Then this might just be fixed by inverting the variable. i.e.

```cpp
lua_getfield(L, index, "on_rightclick"); 
f.can_place_onto = !lua_isfunction(L, -1); 
```

Because, if a node has an `on_rightclick` function, it is **not** placeable onto (this can be bypassed by using Shift+RMB, but that's a different story).

DS-Minetest replied:
There might be other factors which disallow placement without sneaking, eg. formspec meta.
Furthermore, this field has to be sent to the client, too. Because of backwards compatibility you could not send the new bool to the client without inverting it. This will cause the client code to be different from the server code (at least the names).
All in all, it's much simpler to just use keep this as is, a bool field that tells the client whether there's an `on_rightclick` callback defined.

ClobberXD replied:
Hm, how about changing the name to `is_not_placeable` to not require inversion?

DS-Minetest replied:
(The node itself is placeable, so not `is_not_placeable` but `is_not_placeable_to`.)
The name is a bit confusing as you _can_ still place to it. You just have to hold sneak.

ClobberXD replied:
I'll revert back to `rightclickable` for now.

ClobberXD commented:
@Wuzzy2 From your previous comment:

> ... leftclick's main (!) use is “dig/punch”, and rightclick's main use is “use/place”, where with “use” we mean “using the pointed node or thing”. At least this is defined that way in the Esc menu and README file. And it's mostly true. It's only that special items might subvert these definitions.

`repeat_use_place_repetition_interval` :D

On a more serious note, convenience starts to become an issue with long names like this.

DS-Minetest commented:
I suggest using the names `luse` and `ruse`. (The problem here is that "ruse" is an english word.)

ClobberXD commented:
> I suggest using the names `luse` and `ruse`

Well, the whole point of this PR is to get rid of the association between dig and place actions, and `L` and `R` respectively. ;)

DS-Minetest commented:
> > I suggest using the names `luse` and `ruse`
> 
> Well, the whole point of this PR is to get rid of the association between dig and place actions, and `L` and `R` respectively. ;)

Who says that the L and R have to mean left and right? ((Maybe luse is the labor key and ruse is the relevant key.))
Just give the keys very abstract names. But don't define an order, eg. not use1 and use2. Some examples: luse and ruse, iuse and ause, ause and zuse
`luse` and `ruse` would be good because of historical reasons.

ClobberXD commented:
How about `dig_punch` and `place_use`?

DS-Minetest commented:
> How about `dig_punch` and `place_use`?

Then you would have to document eg. that `on_use` is called when the `dig_punch` button is used and not when the `place_use` button is used (which has a "use" in it).
You can't give descriptive names to the buttons, they just do too many different things. (Well you could use endlessly long names like `place_and_secondary_use_and_open_meta_formspec_but_if_sneak_is_pressed_always_place`.)

Wuzzy2 commented:
> (Well you could use endlessly long names like place_and_secondary_use_and_open_meta_formspec_but_if_sneak_is_pressed_always_place.)

:D

It's funny because it's true. Minetest's control scheme is, in fact, a little insane. A simple right-click has just so many meanings.

I still think `place_use` perfectly summarizes it.

> Then you would have to document eg. that on_use is called when the dig_punch button is used and not when the place_use button is used (which has a "use" in it).

That's true, but it just exposes just another insanity of Minetest IMO. It should have never been called `on_use` in the first place since it's clearly for the punch key (LMB). The word “use” in Minetest is used inconsistenly.

Note that the Lua API is problematic in other areas as well. There's a callback literally called `on_rightclick`. This naming will now bite us in the ass as well …

ClobberXD commented:
I agree with Wuzzy, `place_use` covers *most* of the use-cases of that button.

I'm afraid we cant touch the names of the existing stuff that interface with mods (aka the API :P), as that would break backwards compat. I doubt this would even happen for 6.0.0, as a breaking release needn't break all the existing mods unless \*really\* required.

****

I've finally fixed this PR, and I've tested that both

- Binding actions to mouse buttons
- Re-binding dig and place actions to other keys

work properly. Video coming up soon, once I fix my screen recorder.

jastevenson303 commented:
Tested and works as expected.

ClobberXD commented:
Video demonstration: https://youtu.be/0Ca7nEMiq_U

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3012,51 +3008,50 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
 		}
-	} else if (runData.dig_instantly && input->getLeftReleased()) {
-		// Remove e.g. torches faster when clicking instead of holding LMB
+	} else if (runData.dig_instantly && !isKeyDown(KeyType::DIG)) {
```
`!isKeyDown` is perhaps not the same as `getLeftReleased`. The former just tells the state while the latter maybe returns whether the key was released once between the steps.
Or maybe they are the same, I don't know.

paramat replied:
^ Needs consideration.

ClobberXD replied:
Good catch, thanks! It's indeed `wasKeyDown` that should be used here. `isKeyDown` corresponds to `getLeftState`, and `!isKeyDown` isn't the same as `wasKeyDown`.

ClobberXD replied:
I understand that `!isKeyDown` isn't the same as `get(Left|Right)Released`. I'll implement an `isKeyReleased` instead.

ClobberXD replied:
I've added a `wasKeyPressed` which is equivalent to `get(Left|Right)Clicked`. This method is very similar to `wasKeyDown` but unlike the latter, it's state isn't reset immediately after calling but only at the end of `Game::processPlayerInteraction` - this means that this function will return the same value for all `wasKeyPressed` calls in `Game::processPlayerInteraction` in a given iteration.

(Thanks for reminding me to comment on this, @sfan5)

sfan5 replied:
so is `!isKeyDown(KeyType::DIG)` the correct equivalent of `getLeftReleased()` then?

ClobberXD replied:
`getLeftReleased` returns `true` only for one run of `Game::processPlayerInteraction` and is reset (https://github.com/minetest/minetest/pull/7924/files#diff-7f71a92de20504e8a5636b4317ea67b2L3169). So it's indeed different from `!isKeyDown()` which always evaluates to true if the key is up. *But*, the behaviour is the same in this case, as `dig_instantly` and `!isKeyDown` both need to be true for the inner code to run. Once the inner code runs, `dig_instantly` is set to false, and the code never runs again until the key is pressed again. Essentially, `dig_instantly` is already used to keep track of key release events, so `!down` is the same as `released` in this specific context.

ClobberXD replied:
I've changed `wasKeyPressed` to evaluate to true only when the key was pressed down. So this is equivalent to `isKeyDown`, except that it's state would be reset after one step. I've now added a separate `wasKeyReleased` (which I should've done much earlier) which returns true for one step after a key has been released. I'm still in the process of testing this, but this should fix everything (TM).

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3012,51 +3008,50 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
 
-	if (runData.left_punch || input->getLeftClicked())
-		camera->setDigging(0); // left click animation
+	if (runData.punch || isKeyDown(KeyType::DIG))
```
Should `wasKeyDown` be used here?

ClobberXD replied:
Not really, because (hopefully I'm not getting this wrong):

```pseudocode
if (player is punching and they are currently holding down the DIG button)
    set appropriate digging animation
```

DS-Minetest replied:
But
https://github.com/minetest/minetest/blob/5f1cd555cd9d1c64426e173b30b5b792d211c835/src/client/camera.h#L123-L125
says that it starts the animation.
(Also there the buttons are still called left and right.)

ClobberXD replied:
This should indeed be `wasKeyDown`. Changing, thanks for the heads-up.

ClobberXD replied:
Actually, I'll replace this with the new `wasKeyPressed` method...

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3153,7 +3148,7 @@ PointedThing Game::updatePointedThing(
 void Game::handlePointingAtNothing(const ItemStack &playerItem)
 {
-	infostream << "Right Clicked in Air" << std::endl;
+	infostream << "Attempted to place block while pointing at nothing" << std::endl;
```
```suggestion
	infostream << "Attempted to place node while pointing at nothing" << std::endl;
```
or even
```suggestion
	infostream << "Attempted to place item while pointing at nothing" << std::endl;
```

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3193,18 +3188,17 @@ void Game::handlePointingAtNode(const PointedThing &pointed,
-			if (nodedef_manager->get(map.getNode(nodepos)).rightclickable) {
+			// Report place action to server
+			if (nodedef_manager->get(map.getNode(nodepos)).can_place_onto)
 				client->interact(INTERACT_PLACE, pointed);
```
The interact has to be sent to trigger the `on_rightclick` callback. This is not placement. (Well, the action is indeed called "place".)
Anyway, the `can_place_onto` is wrong. It has to be `rightclickable`, as in "has an `on_rightclick` callback".

ClobberXD replied:
I'll just rename it back to `on_rightclick`. I don't want to stall this PR with the ongoing naming contest.

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3245,12 +3240,10 @@ void Game::handlePointingAtNode(const PointedThing &pointed,
 
 				if (def.node_placement_prediction.empty() ||
-						nodedef_manager->get(map.getNode(nodepos)).rightclickable) {
+						nodedef_manager->get(map.getNode(nodepos)).can_place_onto)
```
`rightclickable` is not the same as `can_place_onto`.

ClobberXD replied:
Okay...

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3269,8 +3262,8 @@ bool Game::nodePlacementPrediction(const ItemDefinition &selected_def,
 
-	if (!prediction.empty() && !(nodedef->get(node).rightclickable &&
-			!isKeyDown(KeyType::SNEAK))) {
+	if (!(nodedef->get(node).can_place_onto && !isKeyDown(KeyType::SNEAK)) &&
```
Same as above.

ClobberXD replied:
_Okay..._

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3431,12 +3424,12 @@ void Game::handlePointingAtObject(const PointedThing &pointed,
 		}
 
-		if (input->getLeftClicked())
+		if (isKeyDown(KeyType::DIG))
```
`wasKeyDown`?

ClobberXD replied:
Not sure if that's the way it's supposed to be, but I'll retest after changing this.

ClobberXD replied:
It is indeed `wasKeyDown`. Fixed.

DS-Minetest commented on src/client/game.cpp:
```diff
@@ -3592,7 +3585,7 @@ void Game::handleDigging(const PointedThing &pointed, const v3s16 &nodepos,
 	}
 
-	camera->setDigging(0);  // left click animation
+	camera->setDigging(0);  // Dig animation
```
Another problem with the naming. Digging is also the name of the player model's hand movement.

ClobberXD replied:
Agreed.

\*sigh\*, the naming issues are likely going to stall this PR big-time...

ClobberXD replied:
Ah, are you referring to the function name?

ClobberXD replied:
Thoughts on renaming this to `Camera::setAnim`? Maybe we can also pass an Enum as the parameter, instead of a plain number.

DS-Minetest commented on src/client/inputhandler.cpp:
```diff
@@ -112,56 +114,78 @@ bool MyEventReceiver::OnEvent(const SEvent &event)
 				keyIsDown.set(keyCode);
-				keyWasDown.set(keyCode);
 			} else {
 				keyIsDown.unset(keyCode);
```
Maybe you need a new `keyWasUp`.
```cpp
if (IsKeyDown(keyCode))
	keyWasUp.set(keyCode);
keyIsDown.unset(keyCode);
```

DS-Minetest commented on src/client/inputhandler.cpp:
```diff
@@ -112,56 +114,78 @@ bool MyEventReceiver::OnEvent(const SEvent &event)
+		case EMIE_LMOUSE_PRESSED_DOWN:
+			key = "KEY_LBUTTON";
+			keyIsDown.set(key);
+			keyWasDown.set(key);
```
Not like above?
```cpp
if (!IsKeyDown(key))
	keyWasDown.set(key);
keyIsDown.set(key);
```

DS-Minetest commented on src/client/inputhandler.cpp:
```diff
@@ -112,56 +114,78 @@ bool MyEventReceiver::OnEvent(const SEvent &event)
+		case EMIE_LMOUSE_TRIPLE_CLICK:
+		case EMIE_RMOUSE_TRIPLE_CLICK:
+		case EMIE_MMOUSE_TRIPLE_CLICK:
+		case EMIE_COUNT:
```
Would a `default:` not suffice?

ClobberXD replied:
It should, I just didn't recompile after adding a `default`, which lead me to think that it didn't. Will fix it.

DS-Minetest commented on src/client/inputhandler.cpp:
```diff
@@ -188,38 +212,28 @@ s32 RandomInputHandler::Rand(s32 min, s32 max)
+		{ "keymap_forward", 0.0f, 40 },
+		{ "keymap_left", 0.0f, 40 },
+		{ "keymap_dig", 0.0f, 30 },
+		{ "keymap_place", 0.0f, 15 }
```
```suggestion
		{"keymap_place", 0.0f, 15}
```
https://www.kernel.org/doc/html/latest/process/coding-style.html#spaces
>Do not add spaces around (inside) parenthesized expressions.

ClobberXD replied:
```
(a + b * c / d)  // <--- This is a parenthesized expression

( )  // <--- This is a pair of parenthesis

{  }  // <--- This is a pair of braces

{a, b, c, d}  // <--- This is a... list-like thingy, as far as languages like C++ are concerned

{ a, b, c, d }  // <--- Adding spaces within braces is more readable (esp. with
                //      elements more complex than one-letter identifiers), IMO
```

DS-Minetest commented on src/gui/guiKeyChangeMenu.cpp:
```diff
@@ -290,65 +291,119 @@ bool GUIKeyChangeMenu::resetMenu()
-			//infostream << "Key already in use" << std::endl;
-		}
+	bool shift_went_down = false;
+	if(!shift_down &&
```
```suggestion
	if (!shift_down &&
```

DS-Minetest commented on src/gui/guiKeyChangeMenu.cpp:
```diff
@@ -290,65 +291,119 @@ bool GUIKeyChangeMenu::resetMenu()
-				}
+	// Remove Key already in use message
+	if(this->key_used_text)
+	{
```
`if (...) {`

DS-Minetest commented on src/gui/guiKeyChangeMenu.cpp:
```diff
@@ -290,65 +291,119 @@ bool GUIKeyChangeMenu::resetMenu()
+	}
+	// Display Key already in use message
+	if (std::find(this->key_used.begin(), this->key_used.end(), kp) != this->key_used.end())
+	{
```
Brace.

ClobberXD replied:
... for impact

DS-Minetest commented on src/gui/guiKeyChangeMenu.cpp:
```diff
@@ -290,65 +291,119 @@ bool GUIKeyChangeMenu::resetMenu()
-				return false;
-			}
+		// Allow characters made with shift
+		if(shift_went_down){
```
Spaces.

ClobberXD replied:
```suggestion
		if (shift_went_down) {
```

DS-Minetest commented on src/gui/guiKeyChangeMenu.cpp:
```diff
@@ -290,65 +291,119 @@ bool GUIKeyChangeMenu::resetMenu()
 
-			activeKey = -1;
-			return true;
+bool GUIKeyChangeMenu::OnEvent(const SEvent& event)
```
```suggestion
bool GUIKeyChangeMenu::OnEvent(const SEvent &event)
```

DS-Minetest commented on src/gui/guiKeyChangeMenu.cpp:
```diff
@@ -290,65 +291,119 @@ bool GUIKeyChangeMenu::resetMenu()
+	if ((activeKey >= 0) && (event.EventType == EET_GUI_EVENT) &&
+	    (event.GUIEvent.EventType == gui::EGET_BUTTON_CLICKED)) {
+
+		switch (event.GUIEvent.Caller->getID()) {
```
```cpp
	if ((activeKey >= 0) && (event.EventType == EET_GUI_EVENT) &&
			(event.GUIEvent.EventType == gui::EGET_BUTTON_CLICKED)) {
		switch (event.GUIEvent.Caller->getID()) {
```

DS-Minetest commented on src/gui/guiKeyChangeMenu.cpp:
```diff
@@ -290,65 +291,119 @@ bool GUIKeyChangeMenu::resetMenu()
+
+		switch (event.GUIEvent.Caller->getID()) {
+			case GUI_ID_BACK_BUTTON:
+				break;
```
```cpp
	if ((activeKey >= 0) && (event.EventType == EET_GUI_EVENT) &&
			(event.GUIEvent.EventType == gui::EGET_BUTTON_CLICKED)) {
		switch (event.GUIEvent.Caller->getID()) {
		case GUI_ID_BACK_BUTTON:
			break;
...
```

DS-Minetest commented on src/script/common/c_content.cpp:
```diff
@@ -505,7 +505,7 @@ ContentFeatures read_content_features(lua_State *L, int index)
 
 	lua_getfield(L, index, "on_rightclick");
-	f.rightclickable = lua_isfunction(L, -1);
+	f.can_place_onto = lua_isfunction(L, -1);
```
You can place onto the node even if it has no `on_rightclick`. :wink:

DS-Minetest commented on src/script/common/c_content.cpp:
```diff
@@ -877,8 +877,8 @@ void push_content_features(lua_State *L, const ContentFeatures &c)
-	lua_pushboolean(L, c.rightclickable);
-	lua_setfield(L, -2, "rightclickable");
+	lua_pushboolean(L, c.can_place_onto);
+	lua_setfield(L, -2, "can_place_onto");
```
If you change this, you have to document it in `client_lua_api.txt`.

DS-Minetest commented on src/script/lua_api/l_object.cpp:
```diff
@@ -1418,9 +1418,14 @@ int ObjectRef::l_get_player_control(lua_State *L)
+	lua_setfield(L, -2, "dig");
+	lua_pushboolean(L, control.place);
+	lua_setfield(L, -2, "place");
+	// Legacy fields to ensure mod compatibility
```
Missing documentation.

ClobberXD replied:
Yes, I'm yet to add documentation.

jastevenson303 commented:
Thank you @SmallJoker.  I am curious, how much support is there?  I understand changing a core fundamental such as this can be uncomfortable.

jastevenson303 commented:
@SmallJoker what's the point of rebasing if NO CORE DEV supports this?
http://irc.minetest.net/minetest-hub/2019-09-22#i_5595007

ClobberXD commented:
@rubenwardy supports this, AFAIK

ClobberXD commented:
@rubenwardy _might_ be interested in this PR, as he self-requested a review in #6537 before it was closed.

jastevenson303 commented:
There's a merge conflict now, and I'm not able to resolve it.

This happened before where I used the PR on my own, and I had to stop because I couldn't figure out how to resolve it.  @ClobberXD thank you so much for your contribution, it's really appreciated from here.

It's not really fair to ask so much of contributors.

jastevenson303 commented:
@rubenwardy https://forum.minetest.net/viewtopic.php?p=315289#p315289 this could have gotten into 5.1.  It still could.

Regardless, can you help in supporting this?  Just imagine when it drops, and the on the "New Features" list it shows: * Can now bind mouse buttons.  I think there will be a fair amount of slow clapping.

Imagine all the people who tried Minetest, tried to bind their mouse buttons, and were surprised by what is rather unusual behavior.  Consider DOS games in the early 1990's that allowed binding mouse buttons.  Consider how odd this is, for people who might have a preference.

@sofar @calinou Can I ask for your support?  @ClobberXD has worked to resuscitate a neglected PR, and it stands a chance of happening again.  Minetest has a LOT of features, and a LOT of them are unused, undiscovered, or unimportant.

This is not one of those features.

Calinou commented:
I'm not a core developer, so my approval wouldn't be worth anything really :slightly_frowning_face:

rubenwardy commented:
I would like to see better input support, but I'm feeling too burnt out to review this. Even so, I don't think this should be in 5.1.0 as the deadline is too close now.

I don't see it as an essential feature like you claim. There are tons more features that are way more overdue, such as mobs and a working pathfinder, a main menu that doesn't stink, and particles which actually perform decently

ClobberXD commented:
@jastevenson303: Thanks for the loads of support you've shown towards this PR. I haven't seen the conflict yet, but I'm sure it's easily fixable. Also, I don't have the time to get this merged before 5.1.0, because remember, the PR itself works fine, but the naming problems is the elephant in the room, so to say. While I'd have loved to see this in 5.1.0, I agree with rubenwardy that this isn't *essential* to the release.

jastevenson303 commented:
I was simply stating a possibility.  I would just like to see at least one
core dev support this PR is all.  That way there's at least hope it'll get
merged in 5.2-dev.

@clobberxd it didn't look simple to me!  haha :)

On Wed, Sep 25, 2019 at 4:04 AM ANAND <notifications@github.com> wrote:

> @jastevenson303 <https://github.com/jastevenson303>: Thanks for the loads
> of support you've shown towards this PR. I haven't seen the conflict yet,
> but I'm sure it's easily fixable. Also, I don't have the time to get this
> merged before 5.1.0, because remember, the PR itself works fine, but the
> naming problems is the elephant in the room, so to say. While I'd have
> loved to see this in 5.1.0, I agree with rubenwardy that this isn't
> *essential* to the release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/7924?email_source=notifications&email_token=AFS4UYU6STYVUYTRW4ZS4I3QLMLRPA5CNFSM4GHTYTWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7Q7YKI#issuecomment-534903849>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFS4UYW3UCE6UMLY3KBWYFLQLMLRPANCNFSM4GHTYTWA>
> .
>

paramat commented:
I'm sure many core devs 'support' the concept of this PR, i do too, but 'support' is different to 'approval'.

jastevenson303 commented:
Thank you @paramat.  Do you support mouse binds?  When this PR is rebased
and all issues are addressed, will you approve it?

On Wed, Sep 25, 2019, 1:57 PM Paramat <notifications@github.com> wrote:

> I'm sure many core devs 'support' the concept of this PR, i do too, but
> 'support' is different to 'approval'.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/7924?email_source=notifications&email_token=AFS4UYQ7Q7O3YW3DUT2KI5TQLORCHA5CNFSM4GHTYTWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7SZDRQ#issuecomment-535138758>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFS4UYVLBWMUKAHNX6V5ZALQLORCHANCNFSM4GHTYTWA>
> .
>

ClobberXD commented:
Rebased. Needs testing to ensure that everything still works. I'm yet to find time to address the line comments.

I've also added another commit that fixes *most* of the code-style issues in `src/gui/guiKeyChangeMenu.cpp`

jastevenson303 commented:
Tested and works.

Ferk commented:
About the naming of the actions, to me the right name for the right-button would be "interact".

When holding a The main difference between the "`on_use`" and the (ill-named) "`on_rightclick`" is the fact that `on_right_click` requires a pointed thing. At least the last time I checked (which was quite a while ago).

 - checks lua_api - 

Uhm.. sorry, I think I'm too late. The name was actually changed from `on_rightclick` to `on_secondary_use` already.
Was a change also made to have the action be triggered even when you are pointing at the sky or the nothing?

ClobberXD commented:
> About the naming of the actions, to me the right name for the right-button would be "interact".

*All* actions done by the player falls under "interact", incl. both LMB and RMB actions :)

And yes, `on_secondary_use` is triggered when a player right-clicks, while pointing at nothing. `on_place` is triggered when the player points at an object or a node.

Ferk commented:
> All actions done by the player falls under "interact", incl. both LMB and RMB actions :)

Not necessarily.
You can have "on_use" trigger a change in the item itself without interacting with the environment.
Like for example, when you eat an apple, you are not interacting with anything external.

> And yes, on_secondary_use is triggered when a player right-clicks, while pointing at nothing.

Ok, I'm glad that was changed then. In that case `on_rightclick` is no longer meant to be used for interacting with the world.

ClobberXD commented:
> You can have "on_use" trigger a change in the item itself without interacting with the environment.

Both LMB and RMB actions require the `interact` priv, and are hence considered interaction, at least by Minetest. That's what I meant.

paramat commented:
> When this PR is rebased and all issues are addressed, will you approve it?

Although i support the concept, approving depends on me reviewing and testing and the PR being fit for merge, so i cannot say in advance. I might review but it is not high priority for me.

jastevenson303 commented:
The word is 'activate.'

On Sat, Sep 28, 2019 at 4:31 PM Paramat <notifications@github.com> wrote:

> *@paramat* commented on this pull request.
> ------------------------------
>
> In src/client/game.cpp
> <https://github.com/minetest/minetest/pull/7924#discussion_r329325917>:
>
> > @@ -3012,51 +3008,50 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
>  			client->setCrack(-1, v3s16(0, 0, 0));
>  			runData.dig_time = 0.0;
>  		}
> -	} else if (runData.dig_instantly && input->getLeftReleased()) {
> -		// Remove e.g. torches faster when clicking instead of holding LMB
> +	} else if (runData.dig_instantly && !isKeyDown(KeyType::DIG)) {
>
> ^ Needs consideration.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/7924?email_source=notifications&email_token=AFS4UYWAT7SIUY5ZKULRQDLQL65J5A5CNFSM4GHTYTWKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGIAYFI#discussion_r329325917>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFS4UYUF6SRV5K75ZEA2UNLQL65J5ANCNFSM4GHTYTWA>
> .
>

ClobberXD commented:
Rebased

ClobberXD commented:
The name issue is dragging this PR into oblivion, lol. I'd like to settle on the names "Dig" for LMB and "Place" for RMB, as those are the **primary** functions of the buttons. We'll have to get used to "Use the dig button to attack the monster" and "Use the place button to open/close the door"; we probably would... :P

_Inputs, suggestions, alternatives appreciated!_

Wuzzy2 commented:
I suggest “Punch” for LMB as it's more general-purpose. You can punch nodes, objects and the air. But you can only dig nodes. Therefore, use “Punch”.

I'm fine with “Place” for RMB.

paramat commented:
> "Dig" for LMB and "Place" for RMB, as those are the primary functions of the buttons.

I agree with this, they form a nice pair of opposite actions, and digging and placing are on average the most used functions in MT.

Wuzzy2 commented:
I have no strong feelings towards either alternative.

I prefer “Punch” for LMB and “Place” for RMB.
But I am not opposed to “Dig” for LMB and “Place” for RMB.

ClobberXD commented:
> I have no strong feelings towards either alternative.
> 
> I prefer “Punch” for LMB and “Place” for RMB.
> But I am not opposed to “Dig” for LMB and “Place” for RMB.

While Punch is a great option too, digging is the primary function of the LMB when it comes to Minetest (or so I'd like to think), which is the reason behind choosing that as the name of the button. That, and the fact that the original author had started using that name in a couple of places.

LoneWolfHT commented:
I don't care if it only describes a certain feature as long as I can tell what it does

ClobberXD commented:
I skimmed through the current code to figure out mouse button detection and fix a couple of issues with the PR. Unlike normal keys which have only `isKeyDown` and `wasKeyDown`, mouse buttons have `get(Left|Right)State` (`true` if pressed), `get(Left|Right)Clicked`, and `get(Left|Right)Released`. I'll note down a couple of observations regarding these, in case anyone would like to know.

- `get(Left|Right)State` is directly equivalent to `isKeyDown` for normal keys.
- `get(Left|Right)Clicked` is very similar to `get(Left|Right)State`, but its value is set to false one tick later, at the end of `Game::processPlayerInteraction`. This function is essentially used to check if the button was pressed in the current tick. Used for specifically detecting **clicks**, e.g. left-clicking on a torch to insta-dig it, which is a little faster than when holding down LMB.
- `get(Left|Right)Released` returns true when a mouse button has been released, but its value needs to be manually set to false at the end of `Game::processPlayerInteraction`, just like `get(Left|Right)Clicked`.
- So for most intents and purposes, `wasKeyDown` can be considered as the equivalent to `get(Left|Right)Clicked`.

****

Once I've fixed the existing issues with this PR, I plan on adding support for detecting MMB interactions.

ClobberXD commented:
The merge conflict is scary... *shudders*

ClobberXD commented:
### Update

I'm yet to resolve the merge conflict. But I've added a new function `wasKeyPressed` (short for `wasKeyJustPressed`) which checks whether a key was just pressed. This will be set alongside `isKeyDown`, but will be unset right away in the subsequent iteration of `Game::processPlayerInteraction`. The idea, similar to how `get(Left|Right)Clicked` used to work, is to allow code to detect when the user has just started pressing a key, which is very different from detecting whether a key is down or not. e.g. to begin the wieldhand animation when holding down the dig button - only the initial press needs to be detected to trigger the animation. `isKeyDown` might help detect when a key is pressed, but would continue return true unless and until the key is released, which would result in a bunch of code unnecessarily running over and over again, when they'd have to run only once.

Apart from the merge conflict (I'll try to resolve this myself, but any help would be appreciated), I'd declare the PR ready for testing. Testing instructions in the first post.

LoneWolfHT commented:
If you fix the conflicts I think you'll find people who are willing to test this

ClobberXD commented:
@LoneWolfHT Thanks for the interest. It's the fixing-the-conflicts part that bothers me. There's a lot of conflicts in `guiKeyChangeMenu.cpp`, and unfortunately, I also don't understand much of the code involved. I've been putting this off as it'll take me some time to understand the code and what the original author sought to change here, but I suppose this needs to be done. I'll dig into this today (or within the next couple of days), if possible.

sfan5 commented:
If the key change menu is an issue perhaps make a PR for the functionality first and another one for exposing it in the key change menu?

ClobberXD commented:
That's a very good idea, thanks. Also, this PR won't be incomplete as the keys could be set by editing minetest.conf anyway.

ClobberXD commented:
Removed all changes in `guiKeyChangeMenu.*` and rebased.

- Binding actions to mouse buttons works. :heavy_check_mark: 
- Re-binding dig and place actions to other keys works. :heavy_check_mark: 

I'm very surprised - *nothing* is broken after the rebase. :D

jastevenson303 commented:
Works for me.  I imagine minetest.conf.example will be updated automatically?  Took me a minute to figure out the settings, but it was KEY_RBUTTON and KEY_SPACE.  Thank you!!

ClobberXD commented:
Documentation has been added. CSM method `player:get_control` now returns `dig` and `place` fields instead of `LMB` and `RMB`.

ClobberXD commented on src/script/lua_api/l_localplayer.cpp:
```diff
@@ -231,8 +231,8 @@ int LuaLocalPlayer::l_get_control(lua_State *L)
 	set("zoom", c.zoom);
-	set("LMB", c.LMB);
-	set("RMB", c.RMB);
+	set("dig", c.dig);
+	set("place", c.place);
```
`LMB` and `RMB` are removed from the CSM method entirely. Backwards compat isn't a concern, as this method hasn't been in any release yet.

jastevenson303 commented:
No sorry I was mistaken; I'm not able to eat apples (I did not change LMB/punch bind).

jastevenson303 commented:
...

ClobberXD commented:
@jastevenson303 Try `make clean` and rebuild - worked for me.

ClobberXD commented:
@rubenwardy Thanks for testing!

****

I've re-added a "co-authored-by" attribution to the original author as they had done some much work on this before I adopted it. Please make sure the entire commit message makes it into `master` when merging.

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -6073,14 +6073,37 @@ object you are working with still exists.
-    * example: `{jump=false, right=true, left=false, LMB=false, RMB=false,
-      sneak=true, aux1=false, down=false, up=false}`
+    * The table consists of fields with boolean values representing the pressed
+      keys. Example:
```
Is this example required? Just listing the table keys like before should be enough, isn't it?

ClobberXD replied:
Personally, I think listing the keys would suffice and the example is required at all. Shall I remove it?

ClobberXD replied:
Done

SmallJoker commented on src/client/game.cpp:
```diff
@@ -623,15 +622,15 @@ struct GameRunData {
 	bool dig_instantly;
 	bool digging_blocked;
-	bool left_punch;
+	bool punch;
```
`punching` would be consistent with `digging` - its counterpart.

ClobberXD replied:
You're right, will do.

ClobberXD replied:
Done

jastevenson303 commented:
Unusable until eating and using is fixed.

ClobberXD commented:
@jastevenson303 Thanks for reporting. I too notice some oddities incl. not being able to "use" items.

ClobberXD commented on src/client/game.cpp:
```diff
@@ -3119,55 +3119,56 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
-	if (selected_def.usable && input->getLeftState()) {
-		if (input->getLeftClicked() && (!client->modsLoaded()
-				|| !client->getScript()->on_item_use(selected_item, pointed)))
+	if (selected_def.usable && isKeyDown(KeyType::DIG)) {
+		if (wasKeyPressed(KeyType::DIG) && (!client->modsLoaded() ||
+				!client->getScript()->on_item_use(selected_item, pointed)))
 			client->interact(INTERACT_USE, pointed);
```
This part seems to be the issue with regards to items not being usable. `isKeyDown` returns true, but `wasKeyDown` seems to always return false, which means execution never reaches `client->interact`. `wasKeyPressed` checks if a key was pressed and released (sorta equivalent to a "tap", to make things clear). I think I might need to implement all four keypress states - up, down, pressed, released. :)

ClobberXD replied:
Yeah, I just checked `getLeftClicked` again. It returns true only the instant LMB is pressed, and it's reset at the end of `processPlayerInteraction`. `wasKeyPressed` isn't supposed to replace both `get--Clicked` and `get--Released`...

ClobberXD replied:
`wasKeyReleased` now exists. It's a drop-in replacement for `get(Left|Right)Released`.

ClobberXD commented on src/client/game.cpp:
```diff
@@ -3264,7 +3270,7 @@ PointedThing Game::updatePointedThing(
 void Game::handlePointingAtNothing(const ItemStack &playerItem)
 {
-	infostream << "Right Clicked in Air" << std::endl;
+	infostream << "Attempted to place item while pointing at nothing" << std::endl;
```
This is incorrect, this function handles both LMB and RMB while pointing at nodes.

ClobberXD commented:
Good news, I've fixed everything :tm:. Apart from items being un-use-able, there was also the issue of nodes being placed when the player was rightclicking air before pointing at a node (turned out to be a simple fix, though). Here's what I've tested:

- Digging - test by digging around.
- Item use (LMB) - test using apples or `testfood:*` items in devtest.
- Node placement - test by placing nodes that you dug.
- Item's secondary_use (RMB while pointing at air).
- Node inventory access (on_rightclick?) - test by opening chests in MTG (devtest's chest of everything isn't a node inv. AFAIK).

ClobberXD commented:
@jastevenson303 If you can, please test this PR again and let me know if there's anything else amiss.

jastevenson303 commented:
LOL

![testing](https://user-images.githubusercontent.com/23448162/84178691-d2dd9a80-aa52-11ea-86e3-d8b80fb106a6.png)

tested, FOR WHAT IT'S WORTH... so far so good, but I'll let you know if something odd happens, maybe.

jastevenson303 commented:
Oh I SUPPOSE I should let you know about PLACE_TIME

hint: hold down place and walk around

ClobberXD commented:
@jastevenson303 I don't see any issues with `repeat_place_time` (if that's what you mean). The setting is obeyed correctly as well (I tested this by setting it to `0.1`). Would you be able to elaborate?

jastevenson303 commented:
Nah its cool

On Thu, Jun 11, 2020, 4:31 AM ANAND <notifications@github.com> wrote:

> @jastevenson303 <https://github.com/jastevenson303> I don't see any
> issues with repeat_place_time (if that's what you mean). The setting is
> obeyed correctly as well (I tested this by setting it to 0.1). Would you
> be able to elaborate?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/minetest/minetest/pull/7924#issuecomment-642497903>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFS4UYUMKLASGXWFIHAKHW3RWCI7ZANCNFSM4GHTYTWA>
> .
>

SmallJoker commented on src/defaultsettings.cpp:
```diff
@@ -73,6 +73,8 @@ void set_default_settings(Settings *settings)
 	settings->setDefault("keymap_right", "KEY_KEY_D");
 	settings->setDefault("keymap_jump", "KEY_SPACE");
 	settings->setDefault("keymap_sneak", "KEY_LSHIFT");
+	settings->setDefault("keymap_dig", "KEY_LBUTTON");
+	settings->setDefault("keymap_place", "KEY_RBUTTON");
```
Missing settingtypes.txt entries.

ClobberXD replied:
Aye, will do. o7

ClobberXD replied:
Done

SmallJoker commented on src/client/game.cpp:
```diff
@@ -3119,55 +3123,57 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
 	// Prepare for repeating, unless we're not supposed to
-	if (input->getRightState() && !g_settings->getBool("safe_dig_and_place"))
-		runData.repeat_rightclick_timer += dtime;
+	if (isKeyDown(KeyType::PLACE) && !g_settings->getBool("safe_dig_and_place"))
```
This does not seem to work properly. I've set the following in my configuration:
`keymap_place = KEY_KEY_H` (`H` is not in use anywhere else)

But the nodes are placed two times in a row. Always. No matter how short the key press is.

ClobberXD replied:
Strange, I haven't come across this issue while testing...

ClobberXD replied:
Hmm, this happens only when binding the place button to a key. Investigating...

ClobberXD replied:
Fixed

SmallJoker commented on src/client/game.cpp:
```diff
@@ -3119,55 +3123,57 @@ void Game::processPlayerInteraction(f32 dtime, bool show_hud, bool show_debug)
-		runData.ldown_for_dig = false;
-	}
+	if (!runData.digging && runData.btn_down_for_dig && !isKeyDown(KeyType::DIG))
+		runData.btn_down_for_dig = false;
```
Not working properly either. Released the dig key in between and the crack does not disappear and the progress is kept.

![grafik](https://user-images.githubusercontent.com/1497498/84567801-54973600-ad7b-11ea-92f5-0f3244416a2f.png)

ClobberXD replied:
Not sure if this is related, but I've come across this in master occasionally, especially when I'm trying to dig nodes right at the max range of the tool. I'll look into this, though. Thanks for reporting.

ClobberXD replied:
Good catch, again. This can only be reproduced when binding dig to a normal key. Investigating...

ClobberXD replied:
Fixed.

ClobberXD commented on src/client/inputhandler.cpp:
```diff
@@ -112,56 +114,77 @@ bool MyEventReceiver::OnEvent(const SEvent &event)
+				// In this case, we don't want to let this
+				// stream reach the application as it will cause
+				// certain actions to repeat constantly.
+				if (!IsKeyDown(keyCode)) {
```
I wonder if this check is redundant. Aren't events only sent if there are state changes? Pretty sure a key press event is sent only if the key was just pressed, and isn't continuously sent as long as the key is pressed.

ClobberXD replied:
Apparently not - as long as the key is pressed, one event is sent, and then a stream of events are sent until the key is released; just as this helpful comment suggests. But... why? :thinking:

jastevenson303 replied:
Is this related to an operating system's key repeat settings?  Just curious.

ClobberXD replied:
It seems like it; *shouldn't* be, though...

ClobberXD commented:
@SmallJoker Turns out I hadn't even implemented key-release handlers for the normal keys. :D

Rebased. Fixed all the the things :tm:. PR is ready for review again.

jastevenson303 commented:
Everything appears to be working correctly.  Thank you!

jastevenson303 commented:
I had this really strange experience, as evidenced by the following video:
https://youtu.be/ihmASukVuK4

I have since been unable to reproduce it, and I'm not sure if it's related to setting a player's hand stack (https://github.com/jastevenson303/jastest/blob/master/mods/classes/init.lua#L106) or if it's even related to this PR.

I just wanted to show it just in case, and I'll update if I experience it again.

LoneWolfHT commented:
Can also test the LMB code with this WE command:
```
//lua for i=0, 5, 1 do minetest.after(i, function() minetest.chat_send_all(dump(minetest.get_player_by_name("Lone_Wolf"):get_player_control().LMB))end) end
```

ClobberXD commented:
Thanks for testing this PR, @LoneWolfHT. I'll take a look at the reported issue.

ClobberXD commented:
> When rightclicking with the hammer in my hand it instantly rightclicks the spawned entity instead of requiring a second click

Could you point me to the line of code in your mod that results in this behaviour? Please also provide a minimal reproducible example if possible - that'd be very helpful.

ClobberXD commented on src/network/serverpackethandler.cpp:
```diff
@@ -499,8 +499,7 @@ void Server::process_PlayerPos(RemotePlayer *player, PlayerSAO *playersao,
-	player->control.LMB = (keyPressed & 128);
-	player->control.RMB = (keyPressed & 256);
+	player->control.place = (keyPressed & 256);
 	player->control.zoom = (keyPressed & 512);
```
Aha, server isn't de-serializing the packet properly!

ClobberXD replied:
Fixed

ClobberXD commented:
LMB detection has been fixed.

LoneWolfHT commented:
It's in an on_secondary_use https://github.com/LoneWolfHT/hammer_of_power/blob/master/init.lua#L258
Untested but this code should work. Just drop it in a `test` mod
```lua
minetest.register_node("test:testnode", {
	description = "testnode",
	tiles = {"air.png"},
	groups = {oddly_breakable_by_hand = 3, snappy = 1},
	on_secondary_use = function(_, user, _)
		local name = user:get_player_name()
		local pos = user:get_pos()
		pos.y = pos.y + 1.5

		local obj = minetest.add_entity(
			vector.add(pos, vector.multiply(user:get_look_dir(), 2.5)),
			"test:entity"
		)
	end,
})

minetest.register_entity("test:ent", {
	initial_properties = {
		physical = true,
		collide_with_objects = false,
		selectionbox = {-0.25, -0.25, -0.25, 0.25, 0.25, 0.25},
		pointable = true,
		visual = "cube",
		visual_size = {x = 0.5, y = 0.5, z = 0.5},
		textures = {},
		is_visible = true,
		automatic_rotate = 20,
		glow = 4,
		static_save = false,
	},
	on_punch = function(self, puncher, _, _, dir)
		self.object:remove()
	end,
	on_rightclick = function(self, clicker)
		self.object:remove()
	end,
})
```

SmallJoker commented on doc/lua_api.txt:
```diff
@@ -6105,15 +6105,23 @@ object you are working with still exists.
+    * 6 - sneak
+    * 7 - dig
+    * 8 - place
+    * 9 - zoom
```
`(zoom available since 5.3)`  rebase error.

ClobberXD replied:
Oops, sorry. Will fix.

ClobberXD commented:
I am AFK for a couple of days and I cant test the code you provided until I return. But could you please do me a favour by retesting the same mod after applying the latest updates to the PR?

****

@SmallJoker Thanks for testing. Good to know it works for you. :)

MoNTE48 commented:
Guys, merge this PR after all.
I can’t release PR to improve Android management because I don’t want to make conflicts. This PR will greatly simplify the hack for touchscreen place / dig nodes.

celeron55 commented:
This looks fine enough and the basic feature works for me. We can fix remaining problems with specific mods later, if there are any.

sfan5 commented:
:+1: Not tested but briefly reviewed the code before.

LoneWolfHT commented:
This problem is still present https://github.com/minetest/minetest/pull/7924#pullrequestreview-433633357

jamesalexanderstevenson commented:
You mean the first item, right?

LoneWolfHT commented:
Oh, yeah the second item. It's like all entities are automatically rightclicked when pointed at

jamesalexanderstevenson commented:
Only once, and only after a prior place action is performed successfully on a node.

LoneWolfHT commented:
Someone else is having this issue

LoneWolfHT commented:
I can create an issue linking to the stuff here if wanted

sfan5 commented:
Please create an issue for that, otherwise it will be forgotten.
